File: 2558.Take-Gifts-From-the-Richest-Pile.js
/**
 * @param {number[]} gifts
 * @param {number} k
 * @return {number}
 */
var pickGifts = function (gifts, k) {
  gifts = gifts.map((g) => -1 * g);

  const minHeap = new MinPriorityQueue();

  for (let i = 0; i < gifts.length; i++) {
    minHeap.enqueue(gifts[i]);
  }

  for (let i = 0; i < k; i++) {
    const n = -1 * minHeap.dequeue();
    const sqrt = Math.floor(Math.sqrt(n));
    minHeap.enqueue(-1 * sqrt);
  }

  let sum = 0;

  while (minHeap.size()) {
    const n = -1 * minHeap.dequeue();
    sum += n;
  }

  return sum;
};

File: python/2558.Take-Gifts-From-the-Richest-Pile.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        heap = [-val for val in gifts]
        heapq.heapify(heap)

        while k:
            val = -heapq.heappop(heap)
            heapq.heappush(heap, -floor(sqrt(val)))
            k -= 1

        return -sum(heap)